File: client/Assets/UIDataBinding/databinding/cccomps/CCMyComponent.ts
import { Component } from "cc";
import { ccclass } from "../../convenient";

/**
 * 启用延迟链接
 */
export const EnableLazyAttach = true

@ccclass("CCMyComponent")
export class CCMyComponent extends Component {

	protected static _creatingList: CCMyComponent[] = []
	protected _isCreating: boolean = this.markCreating()
	protected markCreating() {
		CCMyComponent._creatingList.push(this)
		return true
	}

	_deleteAttr(attr: keyof this) {
		delete (this as any)[attr]
	}

	protected onPreDestroy() {

	}

	_onPreDestroy() {
		this._isCreating = false
		this.onPreDestroy()
		super._onPreDestroy && super._onPreDestroy()
	}

	onCreate?() {
		this.onPreload()
	}
	protected __preload() {
		if (this._isCreating) {
			this.onPreload()
		}
	}

	protected onPreload() {

	}

	protected updateAttach() {
		if (EnableLazyAttach) {
			let needAttach = this.needAttach()
			if (needAttach && (!this.isAttached)) {
				this.isAttached = true
				this.onAttach && this.onAttach()
			} else if ((!needAttach) && this.isAttached) {
				this.isAttached = false
				this.onDeattach && this.onDeattach()
			}
		} else {
			if (this.isAttachCalled && (!this.isAttached)) {
				this.isAttached = true
				this.onAttach && this.onAttach()
			} else if ((!this.isAttachCalled) && this.isAttached) {
				this.isAttached = false
				this.onDeattach && this.onDeattach()
			}
		}
	}
	protected needAttach(): boolean {
		return this.isAttachCalled && this.enabledInHierarchy
		// && this.node.parent?.activeInHierarchy
	}
	protected isAttachCalled: boolean = false
	protected isAttached: boolean = false
	protected onEnable() {
		this.updateAttach()
	}
	protected onAfterAttach() {
		this.isAttachCalled = true
		this.updateAttach()
	}
	protected onDisable() {
		this.updateAttach()
	}
	protected onAfterDeattach() {
		this.isAttachCalled = false
		this.updateAttach()
	}

}

File: client/Assets/UIDataBinding/databinding/cccomps/CCMyComponent.cs
using UnityEngine;
using System;
using System.Collections.Generic;

namespace UI.DataBinding
{
	public class CCMyComponent : MonoBehaviour
	{
		/**
		 * 启用延迟链接
		 */
		public static readonly bool EnableLazyAttach = true;

		public CCMyComponent() : base()
		{
			this._isCreating = this.markCreating();
		}

		public virtual CCNodeLife ccNodeLife
		{
			get
			{
				return this.GetComponent<CCNodeLife>();
			}
		}

		protected static List<CCMyComponent> _creatingList = new List<CCMyComponent>();
		protected bool _isCreating = false;
		protected virtual bool markCreating()
		{
			CCMyComponent._creatingList.Add(this);
			return true;
		}

		protected virtual void onPreDestroy()
		{

		}

		protected virtual void OnDestroy()
		{
			this._isCreating = false;
			this.onPreDestroy();
		}

		protected virtual void Awake()
		{
			if (this._isCreating)
			{
				this._isCreating = false;
				this.onPreload();
			}
		}
		public virtual void HandleLoaded()
		{
			if (this._isCreating)
			{
				this._isCreating = false;
				this.onPreload();
			}
		}

		protected virtual void onPreload()
		{

		}

		protected virtual void onAttach()
		{

		}

		protected virtual void onDeattach()
		{

		}

		public virtual void updateAttach()
		{
			if (EnableLazyAttach)
			{
				var needAttach = this.needAttach();
				if (needAttach && (!this.isAttached))
				{
					this.isAttached = true;
					this.onAttach();
				}
				else if ((!needAttach) && this.isAttached)
				{
					this.isAttached = false;
					this.onDeattach();
				}
			}
			else
			{
				if (this.isAttachCalled && (!this.isAttached))
				{
					this.isAttached = true;
					this.onAttach();
				}
				else if ((!this.isAttachCalled) && this.isAttached)
				{
					this.isAttached = false;
					this.onDeattach();
				}
			}
		}

		public virtual bool enabledInHierarchy
		{
			get
			{
				return this.enabled && this.gameObject.activeInHierarchy;
			}
		}
		protected virtual bool needAttach()
		{
			return this.isAttachCalled && this.enabledInHierarchy;
			// && this.node.parent?.activeInHierarchy
		}
		protected bool isAttachCalled = false;
		protected bool isAttached = false;
		protected virtual void onEnable()
		{
			this.updateAttach();
		}
		public virtual void onRequireAttach()
		{
			this.isAttachCalled = true;
			this.updateAttach();
		}
		protected virtual void onDisable()
		{
			this.updateAttach();
		}
		public virtual void onRequireDeattach()
		{
			this.isAttachCalled = false;
			this.updateAttach();
		}

	}
}
